### ISSUE TITLE
Doesn't handle SPF records containing CNAME records

### ISSUE THREAD

**Issue by ad0nis** [securestate/king-phisher#403]
## Issue Description
I experienced an issue when using KP with an SPF record containing a CNAME record which pointed back to a hostname which points back to the same host that KP server is running from. I assume that the fix is to make sure that CNAME records are resolved until they return an A record...

## Reproduction Steps
1. Start King Phisher
2. Configure Campaign
3. Configure SPF record with something like this:
v=spf1 a mx a:cname.domain.com ip4:ip.of.kp.server ~all
4. Attempt to send test email. When the server checks the SPF record, the client will throw the error.

## Environment Details
Host OS: Kali ~2019.2
King Phisher Version: Unsure - Sorry

## Error Details / Stack Trace
```
Error Type: builtins.AttributeError
Error Details: AttributeError("'CNAME' object has no attribute 'address'")
Error UID: 34c11467-aa0f-4cc6-bc35-d6d7ddc969a7
RPC Error: N/A (Not an RPC error)
King Phisher Version: 1.13.1
Platform Version: Linux: Kali kali-rolling kali-rolling
Python Version: 3.7.3
Gtk Version: 3.24.5
Timezone: America/New_York

Thread Information:
===================
 => MainThread (alive=True daemon=False)
    Thread-2 (alive=True daemon=True)
    Thread-1 (alive=True daemon=False)
    Thread-6 (alive=True daemon=True)
    Thread-7 (alive=True daemon=True)

Stack Trace:
============
Traceback (most recent call last):
  File "/usr/share/king-phisher/king_phisher/client/tabs/mail.py", line 325, in signal_button_clicked_sender_start
    if not method():
  File "/usr/share/king-phisher/king_phisher/client/tabs/mail.py", line 259, in _sender_precheck_spf
    spf_result = spf.check_host(spf_test_ip, spf_test_domain, sender=spf_test_sender, timeout=self.config['spf_check_timeout'])
  File "/usr/lib/python3/dist-packages/smoke_zephyr/utilities.py", line 177, in __call__
    result = self._target_function(*args)
  File "/usr/share/king-phisher/king_phisher/spf.py", line 202, in check_host
    return s.check_host()
  File "/usr/share/king-phisher/king_phisher/spf.py", line 284, in check_host
    self.policy = self._check_host(self.ip_address, self.domain, self.sender)
  File "/usr/share/king-phisher/king_phisher/spf.py", line 341, in _check_host
    if self._evaluate_mechanism(ip, domain, sender, directive.mechanism, directive.rvalue):
  File "/usr/share/king-phisher/king_phisher/spf.py", line 427, in _evaluate_mechanism
    if not found and self._hostname_matches_ip(ip, mx_record):
  File "/usr/share/king-phisher/king_phisher/spf.py", line 473, in _hostname_matches_ip
    return str(ip) in tuple(a.address for a in answers)
  File "/usr/share/king-phisher/king_phisher/spf.py", line 473, in <genexpr>
    return str(ip) in tuple(a.address for a in answers)
AttributeError: 'CNAME' object has no attribute 'address'

```

**Comment by zeroSteiner** [comment]
Yup, that looks like a bug. Thanks for reporting this. It might take a little while as I'll have to dig into the RFC to figure out how this should be handled.

**Comment by zeroSteiner** [comment]
I was able to reproduce this, I set up a test environment using the domain name `spf.king-phisher.com`. I'll take this down once testing has completed.

```
python3 tools/spf_check.py -L DEBUG 1.2.3.4 test@spf.king-phisher.com
DEBUG    resolving TXT record for spf.king-phisher.com using nameserver 192.168.1.22 (remaining queries: 9)
DEBUG    parsing 5 directives for domain: spf.king-phisher.com
DEBUG    resolving A   record for spf.king-phisher.com using nameserver 192.168.1.22 (remaining queries: 8)
DEBUG    resolving A   record for spf.king-phisher.com using nameserver 192.168.1.22 resulted in a void lookup
DEBUG    resolving MX  record for spf.king-phisher.com using nameserver 192.168.1.22 (remaining queries: 7)
DEBUG    resolving MX  record for spf.king-phisher.com using nameserver 192.168.1.22 resulted in a void lookup
DEBUG    resolving A   record for cname.spf.king-phisher.com using nameserver 192.168.1.22 (remaining queries: 6)
Traceback (most recent call last):
  File "tools/spf_check.py", line 120, in <module>
    sys.exit(main())
  File "tools/spf_check.py", line 85, in main
    result = spf_test.check_host()
  File "/home/steiner/Repositories/king-phisher/king_phisher/spf.py", line 284, in check_host
    self.policy = self._check_host(self.ip_address, self.domain, self.sender)
  File "/home/steiner/Repositories/king-phisher/king_phisher/spf.py", line 341, in _check_host
    if self._evaluate_mechanism(ip, domain, sender, directive.mechanism, directive.rvalue):
  File "/home/steiner/Repositories/king-phisher/king_phisher/spf.py", line 391, in _evaluate_mechanism
    if self._hostname_matches_ip(ip, rvalue):
  File "/home/steiner/Repositories/king-phisher/king_phisher/spf.py", line 480, in _hostname_matches_ip
    return str(ip) in tuple(a.address for a in answers)
  File "/home/steiner/Repositories/king-phisher/king_phisher/spf.py", line 480, in <genexpr>
    return str(ip) in tuple(a.address for a in answers)
AttributeError: 'CNAME' object has no attribute 'address'
```

**Comment by zeroSteiner** [comment]
Fixed in commit ab91090c2df85f511788dad6a55748694b0d4303. [RFC-7208](https://tools.ietf.org/html/rfc7208#section-5.3) did not clearly specify how CNAME records should be handled while processing the SPF `a` directive. I opted to check the results for `a` record answers and validate with those IP addresses. Should the one we're searching for not be in that group, it'll recursively resolve the `CNAME` records. This is an attempt to avoid invoking a penalty against the DNS query limit.

```
python3 tools/spf_check.py -L DEBUG 11.22.33.44 test@spf.king-phisher.com
DEBUG    resolving TXT record for spf.king-phisher.com using nameserver 192.168.1.22 (remaining queries: 9)
DEBUG    processing spf record: 'v=spf1 a mx a:cname.spf.king-phisher.com ip4:1.2.3.4 ~all'
DEBUG    parsing 5 directives for domain: spf.king-phisher.com
DEBUG    resolving A   record for spf.king-phisher.com using nameserver 192.168.1.22 (remaining queries: 8)
DEBUG    resolving A   record for spf.king-phisher.com using nameserver 192.168.1.22 resulted in a void lookup
DEBUG    resolving MX  record for spf.king-phisher.com using nameserver 192.168.1.22 (remaining queries: 7)
DEBUG    resolving MX  record for spf.king-phisher.com using nameserver 192.168.1.22 resulted in a void lookup
DEBUG    resolving A   record for cname.spf.king-phisher.com using nameserver 192.168.1.22 (remaining queries: 6)
DEBUG    resolving A   record for server.spf.king-phisher.com using nameserver 192.168.1.22 (remaining queries: 5)
DEBUG    top check found matching spf directive: '~all'
[+] spf policy result: softfail
[*] top level spf records found:
[*]   #1 (matched)  spf.king-phisher.com
[*]     #1.1             a
[*]     #1.2             mx
[*]     #1.3             a:cname.spf.king-phisher.com
[*]     #1.4             ip4:1.2.3.4
[*]     #1.5  (matched)  ~all
```

This issue has now been fixed in the dev branch. To obtain the changes, you can either wait for them to be merged into the master branch (which will likely take a few weeks) or you can obtain them immediately by opting into [beta testing](https://github.com/securestate/king-phisher/wiki/Updating-King-Phisher#beta-testing).
